velox/expression/CastExpr-inl.h

```diff
@@ -206,29 +206,109 @@ void CastExpr::applyIntToDecimalCastKernel(
+template <typename FromNativeType, TypeKind ToKind>
+VectorPtr CastExpr::applyDecimalToFloatCast(
     const SelectivityVector& rows,
     const BaseVector& input,
     exec::EvalCtx& context,
-    const TypePtr& fromType) {
+    const TypePtr& fromType,
+    const TypePtr& toType) {
+  using To = typename TypeTraits<ToKind>::NativeType;
+
   VectorPtr result;
-  context.ensureWritable(rows, DOUBLE(), result);
+  context.ensureWritable(rows, toType, result);
   (*result).clearNulls(rows);
-  auto resultBuffer =
-      result->asUnchecked<FlatVector<double>>()->mutableRawValues();
+  auto resultBuffer = result->asUnchecked<FlatVector<To>>()->mutableRawValues();
   const auto precisionScale = getDecimalPrecisionScale(*fromType);
-  const auto simpleInput = input.as<SimpleVector<TInput>>();
+  const auto simpleInput = input.as<SimpleVector<FromNativeType>>();
+  const auto scaleFactor = DecimalUtil::kPowersOfTen[precisionScale.second];
   applyToSelectedNoThrowLocal(context, rows, result, [&](int row) {
-    auto output = util::Converter<TypeKind::DOUBLE, void, false>::cast(
-        simpleInput->valueAt(row));
-    resultBuffer[row] =
-        output / DecimalUtil::kPowersOfTen[precisionScale.second];
+    auto output =
+        util::Converter<ToKind, void, false>::cast(simpleInput->valueAt(row));
+    resultBuffer[row] = output / scaleFactor;
   });
+  return result;
+}
+
+template <typename FromNativeType, TypeKind ToKind>
+VectorPtr CastExpr::applyDecimalToIntegralCast(
+    const SelectivityVector& rows,
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const TypePtr& fromType,
+    const TypePtr& toType) {
+  using To = typename TypeTraits<ToKind>::NativeType;
+
+  VectorPtr result;
+  context.ensureWritable(rows, toType, result);
+  (*result).clearNulls(rows);
+  auto resultBuffer = result->asUnchecked<FlatVector<To>>()->mutableRawValues();
+  const auto precisionScale = getDecimalPrecisionScale(*fromType);
+  const auto simpleInput = input.as<SimpleVector<FromNativeType>>();
+  const auto scaleFactor = DecimalUtil::kPowersOfTen[precisionScale.second];
+  applyToSelectedNoThrowLocal(context, rows, result, [&](int row) {
+    auto value = simpleInput->valueAt(row);
+    auto integralPart = value / scaleFactor;
+    auto fractionPart = value % scaleFactor;
+    auto sign = value >= 0 ? 1 : -1;
+    bool needsRoundUp =
+        (scaleFactor != 1) && (sign * fractionPart >= (scaleFactor >> 1));
+    integralPart += needsRoundUp ? sign : 0;
+    if (integralPart > std::numeric_limits<To>::max() ||
+        integralPart < std::numeric_limits<To>::min()) {
+      if (setNullInResultAtError()) {
+        result->setNull(row, true);
+      } else {
+        context.setVeloxExceptionError(
+            row,
+            makeBadCastException(
+                result->type(),
+                input,
+                row,
+                makeErrorMessage(input, row, toType) + "Out of bounds."));
+        return;
```

Comment:
I wonder why don't we have 'return' statement in the 'if' block? Do we have tests that cover all the code paths here?

Reply:
You are right. I just added a test for that path.

Reply:
For some reason I still see 'return' in 'else' branch, but not in 'if' branch. Should the return be outside of if-else block?

Reply:
Yes, that is by design.
-  If not throwing and setting null, then setting null for all error cases(tested by testDecimalToIntegralCastsOutOfBoundsSetNullOnFailure where all out of bound errors are set into std::nullopt, expecting no exception). 
- In else case, setting up to throw the first exception is enough (tested by testDecimalToIntegralCastsOutOfBounds, where it is expecting to catch the first throw).

Reply:
I'm still confused. Why do we need to set resultBuffer[row] when setNullInResultAtError() is true?

Reply:
My understanding is that applying CAST operator to values equals applying CAST on each, unless it throws. In cases of TRY_CAST like functions(https://learn.microsoft.com/en-us/sql/t-sql/functions/try-convert-transact-sql?view=sql-server-ver16), it does not throw, instead sets corresponding values to null.

In short, TRY_CAST("int", X) where X={'a', '1'} would return {NULL, 1}.  

Do i have misunderstanding?

Reply:
This understanding is correct. 

The 'return' statement in question is for the per-row lambda passed to applyToSelectedNoThrowLocal. Returning from that lambda doesn't prevent next row from being evaluated, but does prevent the code after the 'return' from executing.

Here, I see that we have a single line of code after if-else:

`resultBuffer[row] = static_cast<To>(integralPart);`

I also see that 'return' in the 'else' block prevents this line from executing, but there is no 'return' in the 'if' block to prevent this line from executing. This is confusing to me. I'd expect a 'return' after an if-else block since we don't need to evaluate `resultBuffer[row] = static_cast<To>(integralPart);` in either case.

Reply:
Would you do a couple of experiments: (1) remove 'return' statement; (2) move it after the if-else block. Do you observe any change in behavior ( I expect not. )

Reply:
Ah, i see. You are right, if null bit is set, it does not need to set the value itself.  I have moved 'return' after if-else. 

1. Removing would cause to set exception multiple times in case of X={'a', 'b', 'c'} which is wasteful. If throwing, the recording the first may be enough. Also setting might override exceptions, so will throw exception about 'c', while throwing exception about 'a' is desirable.